champ/lib/encoding/adad.rs

```diff
@@ -0,0 +1,209 @@
+//! Synchronous and asynchronous implementations of ADAD decoders and encoders.
+//!
+//! Following the latest draft of PRC-4: https://pog.network/specification/PIPs/04-ADAD/
+
+use thiserror::Error;
+use tokio::io::{AsyncRead, AsyncReadExt};
+
+#[derive(Error, Debug)]
+pub enum ADADError {
+    #[error("unknown error")]
+    Unknown,
+    #[error("failed to read associated data length")]
+    VarIntReadError(#[from] unsigned_varint::io::ReadError),
+    #[error("failed to read associated data length")]
```

Comment:
Maybe change this msg to distinguish between the one before?

Reply:
oh rip that was suppoesed to be a transparent error